docs/developers/pass_pipeline.md

```diff
@@ -0,0 +1,886 @@
+priority for the consideration of performance.
+
+Besides of the standard tf to hlo pipeline of Mlir-HLO, and the DiscLowerTfPass
+that handles the custom-call ops special for BladeDISC such as RandomUniform and
+TopK, ReviseArgumentsForStaticRankPass is the only pass that needs your
+attention. This is due to that BladeDISC is a dynamic shape but "static rank"
+compiler considering of performance & complexity. Being a static rank compiler
+means that it is still possibly needed to compile more than once when the rank
+of the graph is changing, although this is rarely seen real AI workloads. Here
+is the summary of the differences and similarities of a "static shape compiler"
+and "static rank compiler".
+
+|               | static shape compiler | static rank compiler |
+| -------- | -------- | -------- |
+| requirements  | Any factors that decide the shape should be compile time const | Any factors that decide the rank should be compile time const |
+| inputs to a cluster are categorized into | (1) fixed shape inputs<br>(2) constant inputs | (1) fixed rank inputs<br>(2) fixed shape inputs<br>(3) constant inputs |
+| keys to a compilation cache, or, the infos must be logged during JIT clustering | shapes of (1) & values of (2) | ranks of (1) & shapes of (2) & values of (3) |
+
+If the value or shape of some inputs to a cluster decides the rank of some
+internal tensor, which is analyzed during clustering, the value or shape will to
+be logged into the attributes of the main func, as you can see in the IR. Here
+"cluster" refers to the subgraph that the compiler can support, aka, the cluster
+concept as XLA codebase.
+
+```
+disc.input_shape_3 = dense<0> : tensor<2x2xi64>
+```
+indicates that the originally dynamic shaped %arg3 should be regarded as known
+shape of <2x2xi64> at compile time. "disc-tf-revise-args-for-static-rank" pass
+is responsible to apply the logged attributes onto the IR.
+
+After this pass, we should get the IR mainly made of Mhlo Dialect and Standard
+Dialect, in which the Standard Dialect ops are for shape calculations. This is
+part of the shape calculation and more will be lowered in the following phases. 
+
+
+```python
+!cat tutorial_mhlo.mlir
+```
+
+# 2 HLO graph optimzation & placement passes
+
+The passes in this phase can be devided into a number of categories:
+
+## 2.1 shape-simplifier pass
+
+When insert-tie-shape is set to false, it works as a reentrant pass to propagate
+some known shape information, to eliminate the uneccessary unknown dim sizes. In
+a very simple example:
+
+```
+func @main(%arg0 : tensor<?xf32>, %arg1 : tensor<10xf32>) -> tensor<?xf32> {
+  %0 = tensor.cast %arg1 : tensor<10xf32> to tensor<?xf32>
+  %1 = "mhlo.add"(%arg0, %0) : (tensor<?xf32>, tensor<?xf32>) ->
+  tensor<?xf32> return %1 : tensor<?xf32>
+}
+```
+
+can be optimized into:
+
+```
+func @main(%arg0 : tensor<10xf32>, %arg1 : tensor<10xf32>) -> tensor<10xf32> {
```

Comment:
In my understanding, %arg0 can be optimized as tensor<10xf32> is due to the nature of mhlo.add operation. If my understanding is correct, is it safe to do this optimization based on the function body?